### PR TITLE
Add GeographicMidpointReduction

### PR DESCRIPTION
**Pull request by AlexDBlack**
cc @crockpotveggies

### CODE REVIEW

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Override
+    public IAggregableReduceOp<Writable, List<Writable>> reduceOp() {
+        return new AverageCoordinateReduceOp(delim);
+    }
+
+    @Override
+    public List<String> getColumnsOutputName(String columnInputName) {
+        return null;
```
Comment (crockpotveggies) [inline datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Do we not need to return a new column name? If null, does it just perform the op on the existing column? (Sorry not versed in the new API)

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Override
+    public List<ColumnMetaData> getColumnOutputMetaData(List<String> newColumnName, ColumnMetaData columnInputMeta) {
+        return null;
```
Comment (crockpotveggies) [inline datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Similar question as above.

File: datavec-api/src/test/java/org/datavec/api/transform/reduce/TestReductions.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import static org.junit.Assert.assertEquals;
+
+public class TestReductions {
+
+    @Test
+    public void testGeographicMidPointReduction(){
+
+        //http://www.geomidpoint.com/example.html
+        //That particular example is weighted - have 3x weight for t1, 2x weight for t2, 1x weight for t1
+        Text t1 = new Text("40.7143528,-74.0059731");
```
Comment (crockpotveggies) [inline datavec-api/src/test/java/org/datavec/api/transform/reduce/TestReductions.java]: Maybe move the coordinates to a polar extreme? That will exaggerate any mathematical issues.

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java (hunk 2 above)
Comment (crockpotveggies) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Yea so in testing I'm getting NPEs from the reducer when this is null
```
...java.lang.NullPointerException
  at java.util.ArrayList.addAll(ArrayList.java:577)
  at org.datavec.api.transform.reduce.Reducer.transform(Reducer.java:130)
  at org.datavec.api.transform.sequence.window.ReduceSequenceByWindowTransform.transform(ReduceSequenceByWindowTransform.java:65)
```

Comment (crockpotveggies) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: The code that triggers the NPE
```
val transform = new TransformProcess.Builder(schema)
    .removeAllColumnsExceptFor("Timestamp","MMSI","Lat","Lon")
    .transform(new StringToTimeTransform("Timestamp","dd/MM/YYYY HH:mm:ss",DateTimeZone.UTC))
    .transform(new ConcatenateStringColumns("LatLon", ",", List("Lat","Lon")))
    .convertToSequence("MMSI", new NumericalColumnComparator("Timestamp", true))
    .transform(
        new ReduceSequenceByWindowTransform(
            new Reducer.Builder(ReduceOp.Count).keyColumns("MMSI")
                .countColumns("Timestamp")
                .customReduction("LatLon", new Reductions.GeoAveragingReduction)
                .takeFirstColumns("Timestamp")
                .build(),
            new TimeWindowFunction("Timestamp",10L,TimeUnit.MINUTES)
        )
    )
    .removeAllColumnsExceptFor("LatLon")
    .build
```

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            double z = Math.sin(lat);
+
+            sumx += x;
+            sumy += y;
+            sumz += z;
+            count++;
+        }
+
+        @Override
+        public List<Writable> get() {
```
Comment (huitseeker) [inline datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: This one and the above LGTM.

File: datavec-api/src/test/java/org/datavec/api/transform/reduce/TestReductions.java (hunk 3 above)
Comment (huitseeker) [reply datavec-api/src/test/java/org/datavec/api/transform/reduce/TestReductions.java]: Actually, you may want to test the ka-boom : two polar opposites, or more generally equally-distributed points on the sphere. Their Ka-boom nature should be apparent (i.e. raise an exception).

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            String[] split = str.split(delim);
+            if(split.length != 2){
+                throw new IllegalStateException("Could not parse lat/long string: \"" + str + "\"" );
+            }
+            double latDeg = Double.parseDouble(split[0]);
+            double longDeg = Double.parseDouble(split[1]);
+
+            double lat = latDeg * PI_180;
+            double lng = longDeg * PI_180;
+
```
Comment (huitseeker) [inline datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: The following is nice but costly, I'd like to have an idea of how to bail if we were to deal with infinitesimally close points.

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
+/*
```
Comment (huitseeker) [inline datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: This file seems to not be in the right place. I believe you want

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java (hunk 1 above)
Comment (huitseeker) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]:
```

Comment (AlexDBlack) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Fixed

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java (hunk 2 above)
Comment (AlexDBlack) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Fixed

File: datavec-api/src/test/java/org/datavec/api/transform/reduce/TestReductions.java (hunk 3 above)
Comment (AlexDBlack) [reply datavec-api/src/test/java/org/datavec/api/transform/reduce/TestReductions.java]: Added exception, based on "special case" / pt 9 here:
http://www.geomidpoint.com/calculation.html

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java (hunk 5 above)
Comment (AlexDBlack) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Maybe overkill? But if one of you wants to implement rounding error reduction, by all means go ahead. As to cache of last point: hm, that implies order, which I don't think we can assume in general (like, spark execution, or even just non-sequences).

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java (hunk 6 above)
Comment (AlexDBlack) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Moved to org.datavec.api.transform.reduce.impl.GeographicMidpointReduction

File: datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java (hunk 5 above)
Comment (huitseeker) [reply datavec-api/src/main/java/org/datavec/api/transform/transform/reduce/GeographicMidpointReduction.java]: Yeah, except I *know* @crockpotveggies is going to use this on a list where 50% of points or more are within meters of each other. I guess he'll have to implement the above then :)

**Pull request merged**